state/watcher.go

```diff
@@ -1096,150 +1095,291 @@ func (w *relationUnitsWatcher) loop() (err error) {
+	w := &relationLifeStatusWatcher{
+		commonWatcher:  newCommonWatcher(backend),
+		lifeWatcher:    lifeWatcher,
+		statusWatchers: make(map[int]StringsWatcher),
+		relationKeys:   make(map[string]string),
+		statusOut:      make(chan []string),
+		out:            make(chan []string),
+	}
+	go func() {
+		defer w.finish()
+		w.tomb.Kill(w.loop())
+	}()
+	return w
+}
+
+func (w *relationLifeStatusWatcher) finish() {
+	watcher.Stop(w.lifeWatcher, &w.tomb)
+	for _, sw := range w.statusWatchers {
+		watcher.Stop(sw, &w.tomb)
+	}
+	close(w.out)
+	w.tomb.Done()
+}
+
+func (w *relationLifeStatusWatcher) Changes() <-chan []string {
+	return w.out
+}
+
+type relationLifeDoc struct {
+	DocId string `bson:"_id"`
+	Life  Life   `bson:"life"`
+	Id    int    `bson:"id"`
+}
+
+var relationLifeFields = bson.D{{"_id", 1}, {"life", 1}, {"id", 1}}
+
+// lifeChanged starts or stops status watchers for relations with the
+// specified keys, according to whether the relation is alive or not.
+func (w *relationLifeStatusWatcher) lifeChanged(keys []string) error {
+	relColl, closer := w.db.GetCollection(relationsC)
+	defer closer()
+
+	iter := relColl.Find(bson.D{{"_id", bson.D{{"$in", keys}}}}).Select(relationLifeFields).Iter()
+	var doc relationLifeDoc
+	for iter.Next(&doc) {
+		rid := relationGlobalScope(doc.Id)
+		switch doc.Life {
+		case Alive:
+			members := bson.D{{"_id", rid}}
+			statusFilter := func(id interface{}) bool {
+				k, err := w.backend.strictLocalID(id.(string))
+				if err != nil {
+					return false
+				}
+				return k == rid
+			}
+			sw := newStatusWatcher(w.backend, members, statusFilter, w.statusOut)
+			// Consume initial event
+			<-sw.Changes()
+			w.statusWatchers[doc.Id] = sw
+			w.relationKeys[rid] = w.backend.localID(doc.DocId)
+		default:
+			sw, ok := w.statusWatchers[doc.Id]
+			if ok {
+				watcher.Stop(sw, &w.tomb)
+				delete(w.statusWatchers, doc.Id)
+				delete(w.relationKeys, rid)
+			}
+		}
+	}
+	return iter.Close()
 }
 
-func newRelationStatusWatcher(backend modelBackend,
+func (w *relationLifeStatusWatcher) loop() (err error) {
+	var (
+		sentInitial, ok bool
+		changes         []string
+		out             chan<- []string
+	)
+	for {
+		select {
+		case <-w.watcher.Dead():
+			return stateWatcherDeadError(w.watcher.Err())
+		case <-w.tomb.Dying():
+			return tomb.ErrDying
+		case changes, ok = <-w.lifeWatcher.Changes():
+			// Relation life has changed.
+			if !ok {
+				return watcher.EnsureErr(w.lifeWatcher)
+			}
+			if err = w.lifeChanged(changes); err != nil {
+				return err
+			}
+			if !sentInitial || len(changes) > 0 {
+				out = w.out
+			}
+		case rids, ok := <-w.statusOut:
+			// Relation status has changed.
+			if !ok {
+				return tomb.ErrDying
+			}
+			// Status changes are the global key of the relation.
+			// Look up the relation key.
+			for _, rid := range rids {
+				if key, ok := w.relationKeys[rid]; ok {
+					changes = append(changes, key)
```

Comment:
you should probably use a set.Strings, because this is going to get duplicates (two status changes = two changes in the list; we should coalesce and see only one)

Reply:
yep